hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/traceability/HederaTracer.java

```diff
@@ -133,83 +190,96 @@ private void trackNewAction(final MessageFrame messageFrame, final Consumer<Soli
-                    allActions.add(syntheticInvalidAction);
+                break;
+
+            case REVERT:
+                // deliberate failures do not burn extra gas
+                action.setGasUsed(action.getGas() - frame.getRemainingGas());
+                frame.getRevertReason()
+                        .ifPresentOrElse(
+                                bytes -> action.setRevertReason(bytes.toArrayUnsafe()),
+                                () -> action.setRevertReason(new byte[0]));
+                if (frame.getType().equals(CONTRACT_CREATION)) {
+                    action.setRecipientContract(null);
                 }
-            } else {
-                action.setError(new byte[0]);
-            }
-            if (frame.getType().equals(CONTRACT_CREATION)) {
-                action.setRecipientContract(null);
-            }
+                break;
+
+            case EXCEPTIONAL_HALT, COMPLETED_FAILED:
+                // exceptional exits always burn all gas
+                action.setGasUsed(action.getGas());
+                final var exceptionalHaltReasonOptional = frame.getExceptionalHaltReason();
+                if (exceptionalHaltReasonOptional.isPresent()) {
+                    final var exceptionalHaltReason = exceptionalHaltReasonOptional.get();
+                    action.setError(exceptionalHaltReason.name().getBytes(StandardCharsets.UTF_8));
+                    // when a contract tries to call a non-existing address (resulting in a INVALID_SOLIDITY_ADDRESS
+                    // failure), we have to create a synthetic action recording this, otherwise the details of the
+                    // intended call (e.g. the targeted invalid address) and sequence of events leading to the failure
+                    // are lost
+                    if (action.getCallType().equals(CALL) && exceptionalHaltReason.equals(INVALID_SOLIDITY_ADDRESS)) {
+                        final var syntheticInvalidAction = new SolidityAction(
+                                CALL, frame.getRemainingGas(), null, 0, frame.getMessageStackDepth() + 1);
+                        syntheticInvalidAction.setCallingContract(
+                                EntityId.fromAddress(asMirrorAddress(frame.getContractAddress(), frame)));
+                        syntheticInvalidAction.setTargetedAddress(
+                                Words.toAddress(frame.getStackItem(1)).toArray());
+                        syntheticInvalidAction.setError(
+                                INVALID_SOLIDITY_ADDRESS.name().getBytes(StandardCharsets.UTF_8));
+                        syntheticInvalidAction.setCallOperationType(
+                                toCallOperationType(frame.getCurrentOperation().getOpcode()));
+                        allActions.add(syntheticInvalidAction);
+                    }
+                } else {
+                    action.setError(new byte[0]);
```

Comment:
Do we have any compact summary of when an exceptional halt will fail to set an error?